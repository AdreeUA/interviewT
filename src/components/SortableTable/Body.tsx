import React from 'react'
import * as _ from 'lodash'

interface Props {
  data: any[]
}

interface State {}

class Body extends React.Component<Props, State> {
  render() {
    const { data } = this.props
    return (
      <tbody>
        {!_.isEmpty(data) &&
          data.map(({ fullName, thumbnail, email, phone }) => {
            return (
              <tr key={email}>
                <td>
                  <div className="table__image-wrapper">
                    <img src={thumbnail} className="table__image" alt={fullName} />
                  </div>
                </td>
                <td className="table__name">{fullName}</td>
                <td>{email}</td>
                <td>{phone}</td>
              </tr>
            )
          })}
      </tbody>
    )
  }
}

export default Body
