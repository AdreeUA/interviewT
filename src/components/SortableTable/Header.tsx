import React from 'react'
import { Dispatch, connect } from 'react-redux'
import { StoreState } from '../../store/types'
import { setFilterData, filterUsers } from '../../ducks/users'

interface OwnProps {}

interface ConnectedState {}

interface ConnectedDispatch {
  dispatch: Dispatch<StoreState>
}

type Props = OwnProps & ConnectedState & ConnectedDispatch

interface State {}

const mapStateToProps = (state: StoreState) => ({})

const WrappedHeader = connect<ConnectedState, ConnectedDispatch, OwnProps>(mapStateToProps)(
  class Header extends React.Component<Props, State> {
    tableColumns: { label: string; key: string; sort?: boolean }[] = [
      {
        label: 'Image',
        key: 'thumbnail',
      },
      {
        label: 'Name',
        sort: true,
        key: 'fullName',
      },
      {
        label: 'Email',
        sort: true,
        key: 'email',
      },
      {
        label: 'Phone',
        sort: true,
        key: 'phone',
      },
    ]

    handleChangeSearch = (key: string) => (event: React.SyntheticEvent<any>) => {
      const value = event.currentTarget.value
      this.props.dispatch(setFilterData({ [key]: value }))
      this.props.dispatch(filterUsers())
    }

    render() {
      return (
        <thead>
          <tr>
            {this.tableColumns.map((item) => {
              return (
                <th key={item.key} className={`table__header-${item.key}`}>
                  {item.label}
                  {item.sort && <input className="table__input" onChange={this.handleChangeSearch(item.key)} />}
                </th>
              )
            })}
          </tr>
        </thead>
      )
    }
  },
)

export default WrappedHeader
