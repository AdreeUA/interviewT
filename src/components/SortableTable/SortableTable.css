.sortable-table__table {
  margin-bottom: 15px;
}

.sortable-table__text {
  margin-bottom: 15px;
}
