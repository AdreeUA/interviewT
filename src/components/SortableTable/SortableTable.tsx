import React from 'react'
import { Dispatch, connect } from 'react-redux'
import * as _ from 'lodash'
import { StoreState } from '../../store/types'
import { UserState } from '../../ducks/users'
import Header from './Header'
import Body from './Body'

import './SortableTable.css'

interface OwnProps {}

interface ConnectedState {
  usersFilters: UserState[]
  isRequesting: boolean
}

interface ConnectedDispatch {
  dispatch: Dispatch<StoreState>
}

type Props = OwnProps & ConnectedState & ConnectedDispatch

interface State {}

const mapStateToProps = (state: StoreState) => ({
  usersFilters: state.users.filtersResults,
  isRequesting: state.users.isRequesting,
})

const WrappedFilters = connect<ConnectedState, ConnectedDispatch, OwnProps>(mapStateToProps)(
  class SortableTable extends React.Component<Props, State> {
    render() {
      const data = this.props.usersFilters
      return (
        <div className="sortable-table">
          <table className="table sortable-table__table">
            <Header />
            <Body data={data} />
          </table>
          {_.isEmpty(data) &&
          !this.props.isRequesting && <div className="sortable-table__text text">Empty filters table</div>}
          {this.props.isRequesting && <div className="sortable-table__text text">Loading...</div>}
        </div>
      )
    }
  },
)

export default WrappedFilters
