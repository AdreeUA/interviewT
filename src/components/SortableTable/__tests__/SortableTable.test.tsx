import * as React from 'react'
import thunk from 'redux-thunk'
import { shallow } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import userStateMock, { initialState } from '../../../__mocks__/usersStateMock'
import SortableTable from '../SortableTable'

const middlewares = [ thunk ]
const mockStore = configureMockStore(middlewares)

describe('SortableTable', () => {
  it('render() with initialState', () => {
    const store = mockStore({ users: initialState })
    const wrapper = shallow(<SortableTable />, { context: { store } })
    expect(wrapper.dive()).toMatchSnapshot()
  })
  it('render() with completed state', () => {
    const store = mockStore({ users: userStateMock })
    const wrapper = shallow(<SortableTable />, { context: { store } })
    expect(wrapper.dive()).toMatchSnapshot()
  })
  it('render() with requesting', () => {
    const store = mockStore({ users: { ...userStateMock, isRequesting: true } })
    const wrapper = shallow(<SortableTable />, { context: { store } })
    expect(wrapper.dive()).toMatchSnapshot()
  })
})
