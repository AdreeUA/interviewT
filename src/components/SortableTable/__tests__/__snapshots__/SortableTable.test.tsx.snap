// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortableTable render() with completed state 1`] = `
<div
  className="sortable-table"
>
  <table
    className="table sortable-table__table"
  >
    <Connect(Header) />
    <Body
      data={
        Array [
          Object {
            "email": "ariane.scott@example.com",
            "fullName": "mrs ariane scott",
            "phone": "563-392-7975",
            "thumbnail": "https://randomuser.me/api/portraits/thumb/women/5.jpg",
          },
        ]
      }
    />
  </table>
</div>
`;

exports[`SortableTable render() with initialState 1`] = `
<div
  className="sortable-table"
>
  <table
    className="table sortable-table__table"
  >
    <Connect(Header) />
    <Body
      data={Array []}
    />
  </table>
  <div
    className="sortable-table__text text"
  >
    Empty filters table
  </div>
</div>
`;

exports[`SortableTable render() with requesting 1`] = `
<div
  className="sortable-table"
>
  <table
    className="table sortable-table__table"
  >
    <Connect(Header) />
    <Body
      data={
        Array [
          Object {
            "email": "ariane.scott@example.com",
            "fullName": "mrs ariane scott",
            "phone": "563-392-7975",
            "thumbnail": "https://randomuser.me/api/portraits/thumb/women/5.jpg",
          },
        ]
      }
    />
  </table>
  <div
    className="sortable-table__text text"
  >
    Loading...
  </div>
</div>
`;
